File: SegmentLabeler.cpp
#include "SegmentLabeler.hpp"

Mat labelSegments( Mat segImg ) {
	double minVal;
	double maxVal;
	Point minLoc;

	Mat lblImg;
	segImg.convertTo( lblImg, CV_32F );

	minMaxLoc( lblImg, &minVal, &maxVal );
	// No segments
	if( maxVal == 0 ) {
		return lblImg;
	}

	// Scale to 0-MAX
	lblImg *= pow(2.,32) / maxVal;

	// Label each segment
	minMaxLoc( lblImg, &minVal, &maxVal, &minLoc );
	double nextSegLbl = 0;

	while( minVal == 0 ) {
		nextSegLbl++;
//		cout << "seg lbl = " << nextSegLbl << endl;
		floodFill( lblImg, minLoc, Scalar( nextSegLbl ) );

		minMaxLoc( lblImg, &minVal, &maxVal, &minLoc );

//		namedWindow("seg");
//		imshow("seg", lblImg/maxVal);
//		waitKey(100);
	}

	// Convert the segment pixels to zero
	lblImg.setTo(Scalar(0), segImg);
	// and scale to 0-1
//	lblImg /= nextSegLbl;

	return lblImg;
}

File: segmentlabeler.py
import cv

def labelSegments( segImg ):
	#cv.NamedWindow("seg")
	
	lblImg = cv.CreateMat( segImg.rows, segImg.cols, cv.CV_32F )
	cv.Convert( segImg, lblImg )
	
	
	mml = cv.MinMaxLoc( lblImg )
	# No segments available! All pixels are set to the same type
	if mml[1] == 0:
		cv.Set(lblImg, 1)
		return lblImg
	
	# Scale to 0-MAX
	cv.Scale( lblImg, lblImg, 2.**32/mml[1] )
	
	# Label each segment
	mml = cv.MinMaxLoc( lblImg )
	nextSegLbl = 0.
	while( mml[0] == 0 ):
		nextSegLbl += 1
		cv.FloodFill( lblImg, mml[2], nextSegLbl )
		
		mml = cv.MinMaxLoc( lblImg )
		
		#showImg = lblImg
		#cv.Scale( showImg, showImg, 1./nextSegLbl )
		#print cv.MinMaxLoc( showImg ), '\n'
		#cv.ShowImage("seg", showImg)
		#cv.WaitKey(100)

	cv.Set( lblImg, 0, segImg )
	#cv.DestroyWindow('seg')
	return lblImg
